api/hypershift/v1beta1/hostedcluster_types.go

```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
+	// +optional
+	ExternalNetwork *NetworkParam `json:"externalNetwork,omitempty"`
+
+	// DisableExternalNetwork specifies whether or not to attempt to connect the cluster
+	// to an external network. This allows for the creation of clusters when connecting
+	// to an external network is not possible or desirable, e.g. if using a provider network.
+	// +optional
+	DisableExternalNetwork *bool `json:"disableExternalNetwork,omitempty"`
+
+	// Tags to set on all resources in cluster which support tags
+	// +listType=set
+	// +optional
+	Tags []string `json:"tags,omitempty"`
+}
+
+// OpenStackIdentityReference is a reference to an infrastructure
+// provider identity to be used to provision cluster resources.
+type OpenStackIdentityReference struct {
+	// Name is the name of a secret in the same namespace as the resource being provisioned.
+	// The secret must contain a key named `clouds.yaml` which contains an OpenStack clouds.yaml file.
+	// The secret may optionally contain a key named `cacert` containing a PEM-encoded CA certificate.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// CloudName specifies the name of the entry in the clouds.yaml file to use.
+	// +kubebuilder:validation:Required
+	CloudName string `json:"cloudName"`
+}
+
+type SubnetSpec struct {
+	// DNSNameservers holds a list of DNS server addresses that will be provided when creating
+	// the subnet. These addresses need to have the same IP version as CIDR.
+	DNSNameservers []string `json:"dnsNameservers,omitempty"`
+
+	// AllocationPools is an array of AllocationPool objects that will be applied to OpenStack Subnet being created.
+	// If set, OpenStack will only allocate these IPs for Machines. It will still be possible to create ports from
+	// outside of these ranges manually.
+	AllocationPools []AllocationPool `json:"allocationPools,omitempty"`
+}
+
+type AllocationPool struct {
+	// Start represents the start of the AllocationPool, that is the lowest IP of the pool.
+	// +required
+	Start string `json:"start"`
+
+	// End represents the end of the AlloctionPool, that is the highest IP of the pool.
+	// +required
+	End string `json:"end"`
+}
+
+// RouterParam specifies an OpenStack router to use. It may be specified by either ID or filter, but not both.
+// +kubebuilder:validation:MaxProperties:=1
+// +kubebuilder:validation:MinProperties:=1
+type RouterParam struct {
+	// ID is the ID of the router to use. If ID is provided, the other filters cannot be provided. Must be in UUID format.
+	// +kubebuilder:validation:Format:=uuid
+	// +optional
+	ID *string `json:"id,omitempty"`
+
+	// Filter specifies a filter to select an OpenStack router. If provided, cannot be empty.
+	Filter *RouterFilter `json:"filter,omitempty"`
+}
+
+// RouterFilter specifies a query to select an OpenStack router. At least one property must be set.
+// +kubebuilder:validation:MinProperties:=1
+type RouterFilter struct {
```

Comment:
For each property, can we specifically say whether or not this is required or optional please?